File: notebooksjava/_02Anagram.java
import java.util.Arrays;

public class _02Anagram {

    // 解法1：ソート法 — O(n log n) time, O(n) space
    public boolean isAnagramSort(String s1, String s2) {
        if (s1.length() != s2.length()) return false;
        char[] a = s1.toCharArray();
        char[] b = s2.toCharArray();
        Arrays.sort(a);
        Arrays.sort(b);
        return Arrays.equals(a, b);
    }

    // 解法2：カウント法 — O(n) time, O(1) space（英小文字限定）
    public boolean checkPermutation(String s1, String s2) {
        if (s1.length() != s2.length()) return false;

        int[] counts = new int[26];
        for (char c : s1.toCharArray()) counts[c - 'a']++;
        for (char c : s2.toCharArray()) counts[c - 'a']--;

        for (int count : counts) {
            if (count != 0) return false;
        }
        return true;
    }

    public static void main(String[] args) {
        _02Anagram sol = new _02Anagram();

        // isAnagramSort
        assert sol.isAnagramSort("abc", "cba")   == true  : "abc/cba are anagrams";
        assert sol.isAnagramSort("apple", "pale") == false : "different lengths";
        assert sol.isAnagramSort("listen", "silent") == true;
        assert sol.isAnagramSort("hello", "world")   == false;

        // checkPermutation
        assert sol.checkPermutation("abc", "cba")      == true;
        assert sol.checkPermutation("apple", "pale")   == false;
        assert sol.checkPermutation("listen", "silent") == true;
        assert sol.checkPermutation("hello", "world")   == false;

        System.out.println("All test cases passed ✅");
    }
}

File: notebooksc++/02-anagram.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <cassert>

using namespace std;

// Solution 1: Using sort - O(n log n) time, O(n) space
bool isAnagramSort(string s1, string s2) {
    if (s1.length() != s2.length()) {
        return false;
    }
    
    sort(s1.begin(), s1.end());
    sort(s2.begin(), s2.end());
    
    return s1 == s2;
}

// Solution 2: Using character count array - O(n) time, O(1) space (for lowercase a-z only)
bool checkPermutation(const string& s1, const string& s2) {
    if (s1.length() != s2.length()) {
        return false;
    }
    
    int counts[26] = {0};
    
    for (char c : s1) {
        counts[c - 'a']++;
    }
    
    for (char c : s2) {
        counts[c - 'a']--;
    }
    
    for (int i = 0; i < 26; i++) {
        if (counts[i] != 0) {
            return false;
        }
    }
    
    return true;
}

int main() {
    // Test isAnagramSort
    assert(isAnagramSort("abc", "cba") == true);
    assert(isAnagramSort("apple", "pale") == false);
    assert(isAnagramSort("listen", "silent") == true);
    assert(isAnagramSort("hello", "world") == false);
    
    // Test checkPermutation
    assert(checkPermutation("abc", "cba") == true);
    assert(checkPermutation("apple", "pale") == false);
    assert(checkPermutation("listen", "silent") == true);
    assert(checkPermutation("hello", "world") == false);
    
    cout << "All test cases passed ✅" << endl;
    return 0;
}
